### PR TITLE
Refactor settings manager to allow setting all preferences with env vars and context managers

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This started with looking into #2929 ... and turned into a rabbit hole, but I like where it ended up going.  so putting this here to get some feedback @goanpeca, @sofroniewn, @ppwadhwa ...

I would like for `SettingsManager` to actually _be_ a `BaseSettings` object itself, (given that pydantic has nice support for nested models).  so this PR primarily refactors `SettingsManager` into an evented settings model that has the submodels as fields:

```python
class NapariSettings(BaseNapariSettings):
    application: ApplicationSettings
    appearance: AppearanceSettings
    plugins: PluginsSettings
    shortcuts: ShortcutsSettings
    experimental: ExperimentalSettings
```

so you'll want to compare the new [`napari_settings.py`](https://github.com/napari/napari/pull/2932/files#diff-9fbcb2f349ff81483eb7a65e0e5909b534bcc2129025eb62d1a9bb331e81c297) to the existing [`_manager.py`](https://github.com/napari/napari/pull/2932/files#diff-6a486c032f31eb2d963d430b07f0512056ca90d44fa92f9a287e3e662504c6c8)

it also tries to use pydantic functionality more, when possible... for instance, we're mostly using `settings._sections` instead of `settings.__fields__`, and pydantic has its own `dict(exclude_defaults=True)` method, but we have our own `_remove_default` method and `to_dict` methods.  

This also adds support for nested env var setting, eg...
```sh
NAPARI_APPEARANCE_THEME=light napari
```

... and makes accessing the default values on any evented model easier: `object.defaults` is a property that returns an actual pydantic object... so instead of `settings._defaults['appearance'].theme` you can do `settings.appearance.defaults.theme`.

probably some bugs and omitted functionality left to address, but its more or less working, along with the preferences dialog.

## Type of change

- [x] refactor

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by tlambert03** [comment]
just added a context manager as well:

```python
In [32]: from napari.utils import settings

In [33]: settings.get('appearance.theme')
Out[33]: 'dark'

In [34]: with settings.set({'appearance': {'theme': 'light'}}):
    ...:     print(settings.get('appearance.theme'))
light

In [35]: settings.get('appearance.theme')
Out[35]: 'dark'
```

**Comment by goanpeca** [comment]
Thanks a lot for working on this @tlambert03 

It looks like a great change.

>This also adds support for nested env var setting, eg...
>NAPARI_APPEARANCE_THEME=light napari

We would need to add some ALIAS for `NAPARI_OCTREE` and `NAPARI_ASYNC`?

File: napari/_qt/dialogs/preferences_dialog.py
Comment (goanpeca) [inline napari/_qt/dialogs/preferences_dialog.py]: Is this a temporal comment while tests are fixed?

File: napari/utils/settings/__init__.py
Comment (goanpeca) [inline napari/utils/settings/__init__.py]: We should! but this would be breaking now... (even if new and mostly unused by users?)

File: napari/_qt/dialogs/preferences_dialog.py
Comment (tlambert03) [reply napari/_qt/dialogs/preferences_dialog.py]: yeah... just need to bring back the `_env_settings` or rethink it's access

File: napari/utils/settings/_appearance.py
Comment (goanpeca) [inline napari/utils/settings/_appearance.py]: This does not need to inherit from BaseSettings now?

Comment (tlambert03) [reply napari/utils/settings/_appearance.py]: correct... unless you want each subfield itself to have the capability to save/load, etc...  then only the main `NapariSettings` object would need all that stuff.
If we did someday want to move to a pattern where each sub-field had it's own file, then we could add back the `BaseSettings` inheritance no problem (I did quickly check to make sure...), but for now, might as well inherit from less

**Comment by tlambert03** [comment]
> Thanks a lot for working on this @tlambert03
> 
> It looks like a great change.
> 
> > This also adds support for nested env var setting, eg...
> > NAPARI_APPEARANCE_THEME=light napari
> 
> We would need to add some ALIAS for `NAPARI_OCTREE` and `NAPARI_ASYNC`?

yeah, actually, wanted to ask about that.  since `NAPARI_THEME` also currently, works... are all of the fields in the subsettings available in the top level `NAPARI_` namespace?  seems like we might want to deprecate that in case of future conflicts?

**Review by ppwadhwa — COMMENTED** [review]
It is looking good @tlambert03 !   

I quickly scanned through the code and I think it makes sense to me.  I like that the defaults will be more cleaned up, because that file was starting to get a bit unwieldy.  

I know you said there were probably some bugs and clean up to address, so you may already know about this, but wanted to mention it in case not.  I noticed that I am not able to change the theme or the highlight size from the preferences dialog.  The settings for these do not persist either.

The plugin sorter, however, does let you change the sort order and enabled/disabled setting and persist if I close and re-open napari.

**Comment by goanpeca** [comment]
>yeah, actually, wanted to ask about that. since NAPARI_THEME also currently, works... are all of the fields in the subsettings available in the top level NAPARI_ namespace? seems like we might want to deprecate that in case of future conflicts?

I think at some point we discussed this and decided that they should go without the prefix, so `NAPARI_THEME`, instead of `NAPARI_APPEARANCE_THEME`, so we just need to decide what to do.

**Comment by tlambert03** [comment]
gotcha... yeah, for theme it definitely feels nicer to just use `NAPARI_THEME` instead of `NAPARI_APPEARANCE_THEME`... but it does seem like something we won't be able to do everywhere, which has it's own downsides... @sofroniewn, @jni?

**Comment by goanpeca** [comment]
>gotcha... yeah, for theme it definitely feels nicer to just use NAPARI_THEME instead of NAPARI_APPEARANCE_THEME... but it does seem like something we won't be able to do everywhere, which has it's own downsides...

We might do without the section prefix unless it conflicts, in that case add the prefix?

File: napari/utils/settings/_appearance.py
Comment (goanpeca) [reply napari/utils/settings/_appearance.py]: Understood!

**Comment by tlambert03** [comment]
This is ready for a closer look.  Tests are passing, the settings module is back to full coverage (not quite sure what's going on with codecov there).  

I'll write up some additional docs to add to @goanpeca's current autogenerated docs to cover things like the context manager, and some details around config files.

We should have a discussion about the main access api.  It's a bit confusing at the moment since there is both settings.get and settings.get_settings.  They can be merged, but it will take a little discussion. 

I have some questions about how setting the path with get_settings was supposed to work.  (It seems like it would have to be called by a user before we ever called it internally?  But that seems very easy to break).

Maybe we can discuss at meeting

**Review by sofroniewn — APPROVED** [review]
Ok nice, looks like this is coming along well. I saw a couple ToDo's in comments when I went through, not sure if they are for this PR or for followup ones. Also some discussion around the get methods sound good. I'm going to approve as this is good by me, but I'll wait for after that discussion and @goanpeca and @ppwadhwa to do followup reviews and approve too before merge.

**Review by goanpeca — COMMENTED** [review]
Pretty cool stuff :) 

- So there are some small things missing, the env settings I think?
- Unify set/get so we have hopefully 1 symmetric way to do things (maybe with an optional shorthand as talked on the Dev meeting?)
- Check what happens when multiple settings are created, with respect to things that are connected in a two way fashion, like the theme when changed in the preferences or from the viewer etc.


Let me know when this is ready for another round of review @tlambert03 Thanks!

**Comment by tlambert03** [comment]
I'm gonna continue working on the context stuff in a followup PR, let's get in the base pydantic stuff here.

> So there are some small things missing, the env settings I think?

@goanpeca, can you elaborate what you meant here?

**Comment by goanpeca** [comment]
>@goanpeca, can you elaborate what you meant here?

I meant when using environment variables to define settings and disable the widget in the preferences.

**Comment by sofroniewn** [comment]
Tests are failing now as the import warning is being treated as an error. Might want to add that to an ignore list

```
  E   FutureWarning: 'napari.utils.settings' has moved to 'napari.settings'. This will raise an ImportError in a future version
  =========================== short test summary info ============================
  ERROR  - FutureWarning: 'napari.utils.settings' has moved to 'napari.settings...
  !!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
  =============================== 1 error in 1.43s ===============================
```

Maybe @goanpeca can give this another round of review and then we can get it merged. I know it's been a bit of effort to keep the PR up to date as our other settings and preferences change

**Comment by tlambert03** [comment]
> Tests are failing now as the import warning is being treated as an error. Might want to add that to an ignore list

I'd rather just fix our internal usage... (this is one of the things) I keep needing to update when new PRs come in

**Comment by sofroniewn** [comment]
> I'd rather just fix our internal usage... (this is one of the things) I keep needing to update when new PRs come in

Ah got it - makes sense, I hadn't realized

File: napari/_qt/dialogs/preferences_dialog.py
Comment (tlambert03) [inline napari/_qt/dialogs/preferences_dialog.py:285]: @goanpeca, I ended up deleting this section.

The behavior is a bit different of course (but I think the original intent is met?) tell me what you think.  As I understand it, the goal was to prevent env var settings making it into the "permanent" yaml record.  That will happen with the PR as is, since anything that matches the env var will not be written to disk (see `test_settings_env_variables_do_not_write_to_disk` for details).

i guess I was thinking that it's ok if someone wants to change a setting during a napari session, even if was initiated from an env var... (i.e. this doesn't see env vars as "restricting" the possible values, just temporarily overriding them).
thoughts?

**Comment by sofroniewn** [comment]
@tlambert03 looks like tests are all passing now, is this ready for a final look from @goanpeca and then merge?

**Comment by tlambert03** [comment]
> @tlambert03 looks like tests are all passing now, is this ready for a final look from @goanpeca and then merge?

Yep!

**Comment by sofroniewn** [comment]
Ok, this has sat for quite some time now, all tests are passing and I'd like to merge some other PRs tomorrow might give this conflicts. If I've heard nothing by tomorrow morning I will merge when I wake up. Thanks all!!

**Comment by goanpeca** [comment]
Awesome work @tlambert03 

Quick question:

I just ran `napari --reset` and got

```
napari --reset
Validation errors in config file(s).
The following fields have been reset to the default value:

plugins -> call_order
  none is not an allowed value (type=type_error.none.not_allowed)

Resetting settings to default values.
```

Is that expected 🤔 ?

I got it the first time I did it, so just wondering, the second time I did not get the message.

**Review by goanpeca — APPROVED** [review]
🚀

**Comment by sofroniewn** [comment]
> Is that expected 🤔 ?

I think so, but maybe @tlambert03 can weigh in too. I will merge now though, as if not expected - it sounds like something that could be fixed in a followup PR

**Comment by tlambert03** [comment]
thanks both, 

```
plugins -> call_order
  none is not an allowed value (type=type_error.none.not_allowed)
```

is to be expected with this merge.  The previous declaration was: `call_order: CallOrderDict = Field(None...)` which wasn't strictly valid, since `None` is not a `CallOrderDict` (and `Optional` wasn't specified).  so this PR [gave it a `default_factory` of dict](https://github.com/napari/napari/pull/2932/files#diff-ee3725a89976002c47c63fcdb70170f5029a0c28eb5eb331e6a21bf80115ae50R29-R31) ... hence the correction

**Pull request merged**
